### PR TITLE
GIX-2094: Buy ICP Modal

### PR DESCRIPTION
**Pull request by lmuntaner**
# Motivation

Allow users to easily buy ICPs from the NNS Dapp.

The purchase won't happen in the NNS Dapp though. Instead, the user will be redirected to an external provider with the destination account prefilled with their NNS Dapp main account.

# Changes

* New button in NnsAccountsFooter.
* Adapt Footer to display three buttons.
* New modal BuyIcpModal.
* Add new modal in AccountsModal.
* New mixing for a button with icon.

# Tests

* Add modals within the Accounts route TestIdWrapper. Otherwise, they are not accessible with the page object.
* Test BuyIcpModal
* Test in Accounts route that user can use the new modal.
* Add PO and methods to perform all tests with POs.

# Todos

- [x] Add entry to changelog (if necessary).

### CODE REVIEW

**Comment by lmuntaner** [comment]
Current dependencies on/for this PR:
* main
  * **PR #3796** <a href="https://app.graphite.dev/github/pr/dfinity/nns-dapp/3796?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/dfinity/nns-dapp/3796?utm_source=stack-comment).

**Comment by lmuntaner** [comment]
New UX:

<div class='graphite__hidden'>
          <div>🎥 Video uploaded on Graphite:</div>
            <a href="https://app.graphite.dev/media/video/W99H6e2WWjPgtes6VOJI/539c5361-15b6-4bc5-8241-2cad717d8f1b.mov">
              
            </a>
          </div>
<video src="https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/W99H6e2WWjPgtes6VOJI/539c5361-15b6-4bc5-8241-2cad717d8f1b.mov">Screen Recording 2023-11-20 at 08.52.27.mov</video>

**Comment by lmuntaner** [comment]
@dskloetd @peterpeterparker please review

User story: Use can buy ICPs easily from NNS Dapp.

This implementation: Will redirect to user to a [referral link to Banxa](https://docs.banxa.com/docs/overview#referral-link-url-parametization) with a prefilled destination account.

There is still pending to connect to the actual URL of Banxa. But we're still waiting for some documents and their approval.

This will be the MVP. There are still discussions on what the long term feature will be.

**Comment by dskloetd** [comment]
Who is going to benefit from this referral link?
Is the community clearly informed about this?
Has there been a motion proposal?
I feel very icky about this TBH.

**Comment by lmuntaner** [comment]
> Who is going to benefit from this referral link?
This is the first referral link. One of the upcoming features will be to support multiple integrations. This is the MVP with the simplest integration we found: Banxa.

> Is the community clearly informed about this?
That's under way

> Has there been a motion proposal?
Not yet, afaik

**Comment by dskloetd** [comment]
>>Who is going to benefit from this referral link?
>
>This is the first referral link. One of the upcoming features will be to support multiple integrations. This is the MVP with the simplest integration we found: Banxa.

This doesn't answer my questions. With referral links, there is always somebody receiving commissions for every conversion. Who is going to receive the commissions?

**Comment by lmuntaner** [comment]
> This doesn't answer my questions. With referral links, there is always somebody receiving commissions for every conversion. Who is going to receive the commissions?

Ah, good question. Let me ask that.

**Comment by peterpeterparker** [comment]
➕  Product wise, I absolutely agree with @dskloetd feedback
👨‍🎨 UI wise, the three buttons in the footer is the expectation from Artem?

**Comment by lmuntaner** [comment]
> 👨‍🎨 UI wise, the three buttons in the footer is the expectation from Artem?

Yes, you can see the Figma [here](https://www.figma.com/file/29OdZHbNaOk9RkrfESucjb/NNS-1.03-WIP?type=design&node-id=5156-2831&mode=design&t=D6Pmzq0tVMeMeXWZ-0).

**Comment by peterpeterparker** [comment]
> Yes, you can see the Figma

I see, I guess then there isn't just the feature that I don't like much.

File: frontend/src/lib/modals/accounts/BuyIcpModal.svelte [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+  import { Html, Modal } from "@dfinity/gix-components";
+  import type { Account } from "$lib/types/account";
+  import { i18n } from "$lib/stores/i18n";
+  import IdentifierHash from "$lib/components/ui/IdentifierHash.svelte";
+  import BANXA_LOGO from "$lib/assets/banxa-logo.svg";
+
+  export let account: Account;
+
+  let queryParams: Record<string, string | number>;
+  $: queryParams = {
```
Comment (peterpeterparker) [inline frontend/src/lib/modals/accounts/BuyIcpModal.svelte:11]: not sure this should find place within a component. not sure if it should be a constant or service. wdyt?

File: frontend/src/lib/modals/accounts/BuyIcpModal.svelte [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    blockchain: "BTC",
+    orderMode: "BUY",
+    backgroundColor: "2a1a47",
+    primaryColor: "9b6ef7",
+    secondaryColor: "8b55f6",
+    textColor: "ffffff",
+    walletAddress: account?.identifier,
+  };
+  let url: string;
+  $: url = `https://checkout.banxa.com/?${Object.entries(queryParams)
```
Comment (peterpeterparker) [inline frontend/src/lib/modals/accounts/BuyIcpModal.svelte:26]: likewise for the URL here, should it be an env variable?

File: frontend/src/lib/modals/accounts/BuyIcpModal.svelte [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    <p class="highlight"><Html text={$i18n.accounts.buy_icp_note} /></p>
+
+    <div>
+      <h3>{$i18n.accounts.receiving_icp_address}</h3>
+      <IdentifierHash identifier={account?.identifier} />
+    </div>
+  </div>
+
+  <div class="toolbar">
+    <a
```
Comment (peterpeterparker) [inline frontend/src/lib/modals/accounts/BuyIcpModal.svelte:49]: rel=noreferrer noopener

File: frontend/src/lib/modals/accounts/BuyIcpModal.svelte (hunk 2 above)
Comment (peterpeterparker) [inline frontend/src/lib/modals/accounts/BuyIcpModal.svelte:26]: should the param be encoded? (`encodeURIComponent`)

File: frontend/src/lib/themes/mixins/_button.scss [hunk 4]
```diff
@@ -0,0 +1,6 @@
+@mixin with-icon {
```
Comment (peterpeterparker) [inline frontend/src/lib/themes/mixins/_button.scss:1]: why a mixin and not a style / utils within the gix component?

File: CHANGELOG-Nns-Dapp-unreleased.md [hunk 5]
```diff
@@ -14,6 +14,9 @@ proposal is successful, the changes it released will be moved from this file to
 
 #### Added
 
+* Render withdrawal address on ckBTC burn transactions.
```
Comment (peterpeterparker) [inline CHANGELOG-Nns-Dapp-unreleased.md:17]: That's not part of this PR?

File: frontend/src/lib/components/layout/Footer.svelte [hunk 6]
```diff
@@ -3,9 +3,15 @@
 
   export let testId: string | undefined = undefined;
   export let columns = 2;
+
+  let horizontalSpacing: string = "15%";
```
Comment (peterpeterparker) [inline frontend/src/lib/components/layout/Footer.svelte:7]: Can you solve the styling without JS?

File: frontend/src/lib/modals/accounts/BuyIcpModal.svelte [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+      class="button primary full-width with-icon"
+      href={url}
+      target="_blank"
+      data-tid="buy-icp-banxa-button"
+      ><img
```
Comment (peterpeterparker) [inline frontend/src/lib/modals/accounts/BuyIcpModal.svelte:54]: loading=lazy ?

File: frontend/src/lib/routes/Accounts.svelte [hunk 8]
```diff
@@ -100,13 +100,13 @@
...
-  <CkBTCAccountsModals />
-{:else}
-  <AccountsModals />
-{/if}
+  {#if $isCkBTCUniverseStore}
+    <CkBTCAccountsModals />
+  {:else}
+    <AccountsModals />
+  {/if}
+</TestIdWrapper>
```
Comment (peterpeterparker) [inline frontend/src/lib/routes/Accounts.svelte:109]: not sure why this changed?

**Review by peterpeterparker — APPROVED** [review]
Technically, it looks good to me minus a few minor comments, but this isn't the most extensive review I've ever done.

For the record, I have a few concerns about the feature, and I'm not a fan of the UI/UX. So, I'm strictly approving the code here.

File: CHANGELOG-Nns-Dapp-unreleased.md (hunk 5 above)
Comment (lmuntaner) [reply CHANGELOG-Nns-Dapp-unreleased.md:17]: No, weird. Thanks! Removed.

File: frontend/src/lib/components/layout/Footer.svelte (hunk 6 above)
Comment (lmuntaner) [reply frontend/src/lib/components/layout/Footer.svelte:7]: The number of columns is in JS and the hardcoded style needs to be changed.

I can add a prop to gix-components `Toolbar` to change it there. It's still JS, but I agree that it's nicer.

Not that I added the styling with JS in this PR though...

File: frontend/src/lib/modals/accounts/BuyIcpModal.svelte (hunk 1 above)
Comment (lmuntaner) [reply frontend/src/lib/modals/accounts/BuyIcpModal.svelte:11]: Let me add a todo, yes I agree.

This is pending to fully understand how the integration will work.

File: frontend/src/lib/modals/accounts/BuyIcpModal.svelte (hunk 2 above)
Comment (lmuntaner) [reply frontend/src/lib/modals/accounts/BuyIcpModal.svelte:26]: good point, better be safe.

Comment (lmuntaner) [reply frontend/src/lib/modals/accounts/BuyIcpModal.svelte:26]: Yes, I agree. I'll add a todo and once the integration path is clear I'll improve it.

File: frontend/src/lib/modals/accounts/BuyIcpModal.svelte (hunk 7 above)
Comment (lmuntaner) [reply frontend/src/lib/modals/accounts/BuyIcpModal.svelte:54]: Sure, why not.

File: frontend/src/lib/routes/Accounts.svelte (hunk 8 above)
Comment (lmuntaner) [reply frontend/src/lib/routes/Accounts.svelte:109]: I moved the modals within the `TestIdWrapper` so that the PO of the Accounts.svelte was able to select them.

File: frontend/src/lib/themes/mixins/_button.scss (hunk 4 above)
Comment (lmuntaner) [reply frontend/src/lib/themes/mixins/_button.scss:1]: It can go into gix-components indeed.

I'll a task.
